### ISSUE TITLE
[MNT]: User feedback should use warnings, not logs

### ISSUE THREAD

**Issue by mwaskom** [matplotlib/matplotlib#23422]
### Summary

There are at least two places where matplotlib provides a message to the user about usage patterns that are potentially mistakes using logging functionality, rather than as a warning.

Two examples I've encountered recently:

- [Passing string-formatted numbers as data](https://github.com/matplotlib/matplotlib/blob/v3.5.2/lib/matplotlib/category.py#L234-L237)
- [Passing an RGB tuple to the `c` parameter of `scatter`](https://github.com/matplotlib/matplotlib/blob/v3.5.2/lib/matplotlib/axes/_axes.py#L4203-L4210)

I don't think logging is the right mechanism here:

- The feedback provides very little information about the source of the problem, you just get some text in stderr.
- Related, you can convert a warning to an error and then drop into a debugger to figure out what's happening; I don't think that's the case with logging?
- Warnings can be ignored by third party libraries that are doing something on purpose (i.e., passing numbers-as-strings to use a categorical axis), whereas logs are harder to muffle without intervening.

### Proposed fix

It's possible I am wrong here, but `warnings.warn` feels like the right mechanism for both of these pieces of feedback.

**Comment by story645** [comment]
So agree w/ you and also the discussion in/around #13026 was that the categorical warning shouldn't be a warning since it's the expected/intended behavior and that's why it became info and it maybe shouldn't even exist. [meeting notes about it](https://github.com/matplotlib/ProjectManagement/blob/b81ed87a9cfd7aebce8bbba7a44db7b275384624/meeting_notes/2019/2019.md#7-jan-2019)

**Comment by mwaskom** [comment]
Ah thanks missed that original discussion.

It seems like @jklymak made the opposite argument from point 3 here, that logging is preferable because it's easier for knowledgable users to disable it. I'm not sure I understand that argument? Personally I find Python's logging interface confounding so I'm not sure I share the intuition. But my concern here is primarily about 3rd party libraries, who really shouldn't be messing with matplotlib's logging settings IMO.

It also feels odd to me to be logging to communicate with a user in the moment ... logging feels more about communicating with a user in the future. But that's maybe just a personal inclination.

(I'd also prefer if the categorical warning were simply removed, although I appreciate that it's a source of confusion and gather that the ship has sailed).

**Comment by jklymak** [comment]
>  But my concern here is primarily about 3rd party libraries, who really shouldn't be messing with matplotlib's logging settings IMO.

I am by no means an expert in deciding on the best way to pass info between libraries.  But I'm not clear how messing with Matplotlib's logging is any worse than messing with Matplotlib's warnings?    https://docs.python.org/3/howto/logging.html is the basic guidance.  

For the categorical info message, that really seems to not be something the user should see unless they are debugging.  I think its a little more urgent than "debug", which barfs out a lot of output, but not as urgent as a "warning".  For the default logging config, users should not see this message at all, only if they are in the habit of bumping their log level to see what is breaking.  

For the RGB message, that seems something a user is more likely to get wrong than a library, in which case the log message seems the best place to put this according to my parsing of the guidelines?  

But I don't feel strongly about any of this, and I am not speaking from a wealth of experience dealing with maintaining a downstream library...

**Comment by mwaskom** [comment]
Well the warnings module has a context manager for temporarily changing warning filters. I’m not aware of anything similar for logging?

**Comment by mwaskom** [comment]
> For the RGB message, that seems something a user is more likely to get wrong than a library, in which case the log message seems the best place to put this according to my parsing of the guidelines?

Maybe, although the proximate cause for this issue was someone opening a seaborn issue about this log that I was (eventually) able to trace down to a problem in pandas. Which would have been a lot easier if I could have just flipped a switch to make the warning an error and get a stack trace. I agree that in general that `c=rgb_tuple` is not something a library _should_ be doing on purpose.

**Comment by mwaskom** [comment]
Another case where it is useful to turn warnings into errors is in test suites.

**Comment by story645** [comment]
> For the categorical issue, warnings show by default and you don't want warnings to show up if the user is doing something they indeed mean to be doing.

this was always the problem w/ this warning though, right? that it's the totally expected behavior of the library and also something many users may not have intended. And the users who may not have intended it and therefore may benefit most from seeing the warning, are maybe least likely to see the warning since it's an info log.  Just raising that I think there's maybe a UX issue here, though also wondering if at this point it can be removed given categoricals went in in 2.1

**Comment by mwaskom** [comment]
I think you can use warnings and satisfy your desire for a `plt.helpful` mode by defining a custom warning class (`MatplotlibSharpEdgeWarning`) and setting filters on it.

I realized why I am seeing the categorical log even when not asking for it; some transitive dependency does `logging.basicConfig(“INFO”)` which activates a lot more verbosity I didn’t want :(

**Comment by timhoffm** [comment]
From https://docs.python.org/3/howto/logging.html#when-to-use-logging

The best tool for the task:

- [warnings.warn()](https://docs.python.org/3/library/warnings.html#warnings.warn) in library code if the issue is avoidable and the client application should be modified to eliminate the warning
- [logging.warning()](https://docs.python.org/3/library/logging.html#logging.warning) if there is nothing the client application can do about the situation, but the event should still be noted

Applied to the cases:
- [Passing string-formatted numbers as data](https://github.com/matplotlib/matplotlib/blob/v3.5.2/lib/matplotlib/category.py#L234-L237)
  It is a valid use case, but most of the time not what the user intended. One can argue both ways:
  - We don't want false positives -> do not show the warning by default -> `logging`
  - We want to show the warning by default to warn most users -> `warnings.warn`, and we accept some false positives, which these users could filter out.
  I'm now inclined towards the second way.
- [Passing an RGB tuple to the `c` parameter of `scatter`](https://github.com/matplotlib/matplotlib/blob/v3.5.2/lib/matplotlib/axes/_axes.py#L4203-L4210)
  There are unambiguous ways for the user to specify the color (see the message). -> User should take action -> We should use `warnings.warn()`

**Comment by jklymak** [comment]
The rgb case I agree should just be a warning

Still not convinced about categorical.  The point was if someone does something and it behaves unexpectedly we can ask them to turn logging on. If someone does something and it works, and is legitimate, I don't think they should have to filter a warning.

**Comment by timhoffm** [comment]
Usually I'd argee with you that legitimate use cases should not issue warnings.

But weighting the likelihood, we may want to deviate from this:
I assume, the misuse (accidentally passing strings but intending numbers) is far more likely than the legitimate use (wants string labels that happen to be numbers).

If we strictly follow the above rule, we make life much harder for the majority of users. You say "The point was if someone does something and it behaves unexpectedly we can ask them to turn logging on." That's very far down the line from the problem: The user sees a strange output. Then search for a solution. Maybe they find  [our FAQ](https://matplotlib.org/stable/users/faq/howto_faq.html#why-do-i-have-so-many-ticks-and-or-why-are-they-out-of-order) or they ask on discourse or open an issue. It would be much more efficient for them if we would issue a warning right away.

OTOH I'd argue that number labels is a special enough case that we can burden users to explicitly consent to this. - They may just live with the warning or, suppress it with a filter (which we should describe in the warning message).

**Comment by mwaskom** [comment]
> The point was if someone does something and it behaves unexpectedly we can ask them to turn logging on.

I still think logging is the wrong mechanism here, but this would suggest `DEBUG` and not `INFO` is the right level.

> one could make this dependent on the plot type

This has some appeal and logic seems relatively sound, but I would expect that this might get complicated and difficult to maintain. One other distinction that you could draw is between calling a plotting method vs. calling `Axis.update_units` directly. If the user is doing the latter, I think it's safe to say they know what they're doing and really do want  numbers-as-strings.

**Comment by jklymak** [comment]
> I still think logging is the wrong mechanism here, but this would suggest DEBUG and not INFO is the right level.

Well the nice thing about INFO is that it is not super cluttered, whereas DEBUG is very verbose.  Again, the default level is WARNING, so most users shouldn't see this log message.  Asking folks who have a mysterious problem to turn on logging to INFO should give them something useful.  DEBUG is largely for the devs.  

> Usually I'd argee with you that legitimate use cases should not issue warnings.

The original problem is numbers (or dates) being stored in a list or object array as strings.   Unfortunately, many csv file decoders will return these as text if you don't specify the dtype, and a _lot_ of people use csv files.  

If we really think that 99.9% of the users who pass `plot(['1', '2', '3'])` mean for that to be a float array, such that we are willing to issue a warning to the 0.1% who mean for those to be categories, maybe we should go back to the original idea of just converting the array to float (or `datetime64`) and providing a mechanism to force the units to be categorical.

**Comment by pedrompecanha** [comment]
Hello everyone! I see that the second example was marked was done. What about the first one? Could it still be changed to a warning? Thanks!

**Comment by timhoffm** [comment]
> If we really think that 99.9% of the users who pass plot(['1', '2', '3']) mean for that to be a float array, such that we are willing to issue a warning to the 0.1% who mean for those to be categories, maybe we should go back to the original idea of just converting the array to float (or datetime64) and providing a mechanism to force the units to be categorical.

No. We don't want to support sloppy programming. If the user wants numbers, they should provide numbers. The warning has the purpose to make the user aware that they should likely convert their data to numbers.

There are exactly two possible cases
a) The user wanted numbers but accidentally passed strings - in this case, we should `warnings.warn`
b) The user actually wanted numeric  categorical labels - in this case we should do nothing

Unfortunately, we cannot infer the user intention reliably. This leaves us with three options
1) Change to `warnings.warn` under the assumtion that a) is much more likely, and accepting a small false-positive rate of warning for b)
2) keep `logging.warning` with the argument that the case b) users do not feel obliged to act on it (`logging` is only for information, whereas `warnings` imply a request for change of user code)
3) Do not warn at all because we fundamentally don't want a false-positive warnings for b)

Personally, I'm +0.75 for 1)

**Comment by jklymak** [comment]
I don't think we should _warn_ on potentially valid input.

**Comment by anntzer** [comment]
I emphatically agree that potentially valid input should never warn.

**Comment by mwaskom** [comment]
What is that logging call supposed to achieve that is different from “warning” then?

**Comment by jklymak** [comment]
Logging is opt-in and most people leave it off. If someone has a mysterious bug they can turn it on and maybe get some info.  A warning would need to be opt-out.

**Comment by mwaskom** [comment]
To circle back to the beginning of the thread, Python logging is “opt in” in a very crude sense and a line of code somewhere in your dependency tree can opt you into matplotlib logging without you wanting that (how I ran into this).

This still just feels like the wrong mechanism for this kind of information but if you are against it being a warning (I’m fine with that) and still think it adds value (I agree this is a common gotcha, but I am skeptical anyone has thought to debug it by saying “I know, I will turn on matplotlib logging”) we can go ahead and close the thread.

**Comment by timhoffm** [comment]
I think we all agree tha

- if there was only the unintended strings case, an immediate warning (I.e. warnings.warn) is more helpful than an opt-in (logging).
- valid input should never warn.

The decision here is a trade-off. The intended all string numbers case should be quite rare. It's a judgement call: How likely is the valid use case vs. the mis-use? 1%, 0.1%? And is there a limit where we say it is acceptable to show a false positive warning in rare cases to help with a common pitfall.

**Comment by mwaskom** [comment]
To be clear, in the seaborn usecase, it is 100% intentional to be sending in string-typed numbers.

**Comment by timhoffm** [comment]
@mwaskom Do I understand you correctly? You intentionally send string-tryped numbers. But you still want us to warnings.warn on them? Do you plan to filter the warning?

**Comment by mwaskom** [comment]
Right, it would be easier for me overall if there were just no feedback here, though I am sympathetic about this being a common gotcha. But whereas with `warnings.catch_warnings`/`warnings.simplefilter` there is an established pattern for ignoring "expected" warnings, I am not aware of an analogous pattern for ignoring logs. That's not to say it's impossible, but something feels wrong about one library manipulating the warning settings of another library.

Maybe I am just not used to normal patterns because it's pretty unusual for libraries in the scientific Python ecosystem to use logging; logging feels like something that is usually done at the application layer, not the library layer. But that goes back to why it feels like matplotlib is doing something odd here.

**Comment by mwaskom** [comment]
As a maybe orthogonal conversation, maybe `axis.update_units` could have a "verbosity" parameter that sets some state to prevent messages (logging or warning) of this kind. Would need some additional thinking though.

**Comment by jklymak** [comment]
> but something feels wrong about one library manipulating the warning settings of another library.

Matplotlib is not manipulating any other library's settings.  I'm not sure what you mean by this?  In normal usage of Matplotlib there are _no_ logging messages.  If you turn logging on, then yes, you will see our messages, but it's hard to see why that is something we are doing that is incorrect.

**Comment by ksunden** [comment]
@mwaskom was speaking from the perspective of `seaborn` modifying `matplotlib`s log settings for the case where numeric strings are added intentionally to avoid warnings from being printed. (though such warnings only appear with non-default settings, so unclear to me if that is necessary)

**Comment by mwaskom** [comment]
> was speaking from the perspective of seaborn modifying matplotlibs log settings for the case where numeric strings are added intentionally to avoid warnings from being printed.

that's right, thanks

> (though such warnings only appear with non-default settings, so unclear to me if that is necessary)

i am raising this both as the maintainer of seaborn and as a user who sometimes works in a codebase where a transitive dependency does `logging.basicConfig(level=INFO)`, which is why i end up seeing these messages (i do think it's telling that you call them "warnings" 🙂). so this is an example of a case where i did not personally opt-into seeing the logs, and i am also not (directly) doing the thing that is triggering them, resulting in a bunch of confusing and ultimately meaningless warnings cluttering my notebook. and even though i'm a knowledgable user of matplotlib and also wrote the relevant seaborn code, tracking down the source of them was quite difficult because (unlike with a warning) the log was not associated with any specific call into matplotlib. which, all together, feels like bad ux.

**Comment by jklymak** [comment]
> who sometimes works in a codebase where a transitive dependency does logging.basicConfig(level=INFO)

Libraries should never do that - I think your major complaint is with that library, not Matplotlib.  Users who set the logging level should expect to see a bunch of messages.  However, libraries should never set the logging level for users.    

If, as a library maintainer, you really want to suppress _our_ logging message, you can temporarily reset the logging level when you call us:

```
mlog = logging.get_logger('matplotlib')
level = mlog.getEffectiveLevel()
mlog.setLevel(logging.CRITICAL)
ax.plot_matplotlibthing()
mlog.setLevel(level)
```
If you are doing this a lot, you could easily write a context manager.   Its a bit of a pain for you as a maintainer, however, in my opinion it is much more burdensome for end-users to have to filter warnings for valid code than for a maintainer to suppress log messages they don't want.

**Comment by mwaskom** [comment]
It feels like there are a whole additional set of arguments here about why the logging approach is unusual and provides bad ux that you’re not engaging with.

**Comment by jklymak** [comment]
> It feels like there are a whole additional set of arguments here about why the logging approach is unusual and provides bad ux that you’re not engaging with.

I think that's true, because the basic argument is that we are spamming your console.  But we are not, your dependency is.  

Matplotlib has _always_ had logging information being spat out if you set a level of verbosity.  We turned those messages into standard logging calls quite a few years ago because it is a standard library with standard ways of controlling the level of output.  If the argument is general, in that we should never use `logging` to provide extra debugging info to users, then we would have to go back to conditional `print` statements, which seems a huge step backwards, or not provide extra debugging info.  

If the argument is specific to this situation, as @timhoffm says it's a judgement call if we warn for every time a user does this, or if we just log.   I personally don't think we should warn users on valid inputs.  I guess we could argue about the level of the message, but INFO is more verbose than WARNING.  
 
> I think very good suggestion to make a CustomWarning

This is still a warning that users will have to explicitly filter even if they are using the library properly.  

The [logging tutorial](https://docs.python.org/3/howto/logging.html#logging-basic-tutorial) suggests that `warnings.warn` should be used "in library code if the issue is avoidable and the client application should be modified to eliminate the warning".  That doesn't apply to this case if the user means to do `plot(["1", "3", "2"], [1, 2, 3])` - there is no code the user could change to eliminate the warning, no matter what it is called.

**Comment by mwaskom** [comment]
> I think that's true, because the basic argument is that we are spamming your console. But we are not, your dependency is.

To be pedantic, you're spamming my *notebook*. The distinction actually matters because notebooks with matplotlib plots are often (in a data science context) intended as communicative documents so there is extra cost to pointless content in stderr.

Beyond that, I've made a number of other points about why logging is less effective for debugging (from my OP):

> - The feedback provides very little information about the source of the problem, you just get some text in stderr.
> - Related, you can convert a warning to an error and then drop into a debugger to figure out what's happening; I don't think that's the case with logging?
> - Warnings can be intercepted by third party libraries that are doing something on purpose (i.e., passing numbers-as-strings to use a categorical axis), whereas logs are harder to muffle without intervening.

Plus (in most cases) you need to know to switch on INFO mode logging, which is an unusual pattern in scientific Python — logging isn't to my knowledge used in the other core libraries.

> Its a bit of a pain for you as a maintainer, however, in my opinion it is much more burdensome for end-users to have to filter warnings for valid code than for a maintainer to suppress log messages they don't want.

And then circling back here, I don't really think this distinction is relevant. Set aside my poorly-behaved dependency opting me into matplotlib logging. If I, as a user, wanted *useful* information from matplotlib and activated its logging, I'd still need the third party libraries like seaborn to do the hacky manipulation of matplotlib's log level, otherwise I would be getting uninformative "warnings" for valid code.

**Comment by rcomer** [comment]
> it's questionable whether logging helps anyone. This is a rough guess based on my behavior, but I think most folks are more likely to Google strange behavior than to think to check logs they may not even know about.

I didn’t know we did logging until I reviewed #26802…

**Comment by mwaskom** [comment]
I'm not exactly sure how it would work, but it also seems possible to have matplotlib's "maybe warnings" suppressed by default and able to be toggled on with a `plt.verbose_warnings()` (or whatever name) switch.

**Comment by timhoffm** [comment]
I suggest to move from `_log.info()` to `_log.warn()` for the remaining case

- [Passing string-formatted numbers as data](https://github.com/matplotlib/matplotlib/blob/v3.5.2/lib/matplotlib/category.py#L234-L237)

As discussed above `warnings.warn` is only for cases that the user can do something about. But there can be valid cases for "numbers as categorical data". This forces us to stay with log. However, the default log level in the logging framework is warning. So, if the user/app does not explicitly increase the log level, the log.info here is lost.
